### ISSUE TITLE
Add SIMD operations that use f16 and f128

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust#125440]
Eventually we will want to be able to make use of simd operations for f16 and f128, now that we have primitives to represent them. Possibilities that I know of:

- Aarch64 neon supports `float16x{4,8,16}` https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[Neon]&f:@navigationhierarchiesreturnbasetype=[float]&f:@navigationhierarchieselementbitsize=[16]&q=.
- Arm sve supports `float16x{1,2}` https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiesreturnbasetype=[float]&f:@navigationhierarchieselementbitsize=[16]&f:@navigationhierarchiessimdisa=[sve2,sve]&q=
- RISC-V apparently has both f16 and f128 https://five-embeddev.com/riscv-user-isa-manual/riscv-user-2.2/v.html

Probably some work/research overlap with adding assembly https://github.com/rust-lang/rust/issues/125398

**Comment by kjetilkjeka** [comment]
Nvidia ptx [(`--target nvptx64-nvidia-cuda`)](https://doc.rust-lang.org/nightly/rustc/platform-support/nvptx64-nvidia-cuda.html) also support [arithmetic instructions for `f16` and `f16x2` SIMD](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions). 

Making this work is an important step for making the ptx target "feature complete" with languages traditionally used for GPGPU. Let me know if there's anything I can do to support this.

**Comment by tgross35** [comment]
Thanks, I'll add that to the top list.

It looks like it might not be too hard to add new simd intrinsics on that platform? I have no clue but https://github.com/rust-lang/stdarch/blob/df3618d9f35165f4bc548114e511c49c29e1fd9b/crates/core_arch/src/nvptx/mod.rs is pretty straightforward if you want to give it a shot at some point

**Comment by kjetilkjeka** [comment]
I just tested `f16` on nvptx now and I don't think I realized how many of the pieces was already put together. That's great!

I looked a bit around in SIMD instructions for other arches and I think this is, as you say, pretty straightforward. I will give it a shot. Hopefully I will get around to creating a PR next week.

**Comment by kjetilkjeka** [comment]
Took me a bit longer than I originally hoped for but I ended up creating a PR for (most) nvptx `f16x2` intrinsics and getting it merged. https://github.com/rust-lang/stdarch/pull/1626

I have also noticed that we're lacking `portable_simd` variants of `f16` and it's not being tracked by this issue. Is that outside the scope of this issue or just not added yet? Is anyone already coordinating with the portable_simd project or is it simply being blocked by other features that needs to land first?

**Comment by tgross35** [comment]
> Took me a bit longer than I originally hoped for but I ended up creating a PR for (most) nvptx `f16x2` intrinsics and getting it merged. [rust-lang/stdarch#1626](https://github.com/rust-lang/stdarch/pull/1626)

Awesome news, thanks for the update! Looks like there is an open PR to get the new changes https://github.com/rust-lang/rust/pull/128866.

> I have also noticed that we're lacking `portable_simd` variants of `f16` and it's not being tracked by this issue. Is that outside the scope of this issue or just not added yet? Is anyone already coordinating with the portable_simd project or is it simply being blocked by other features that needs to land first?

I'll add it to the issue, no particular reason outside of being lower priority than the intrinsics.

**Comment by michael-p** [comment]
Another datapoint: [ARM Helium](https://www.arm.com/technologies/helium) (a.k.a. Cortex-M Vector Extension) also support SIMD instructions for f16 but not for f128: https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[%5BNeon%5D,Neon,Helium]&f:@navigationhierarchiesreturnbasetype=[%5Bfloat%5D]&f:@navigationhierarchieselementbitsize=[%5B16%5D]&q=

**Comment by krishpranav** [comment]
i will try to work on this issue.

**Comment by krishpranav** [comment]
@tgross35: https://github.com/rust-lang/stdarch/pull/1701

**Comment by LorrensP-2158466** [comment]
Hi, can I help with this issue? I have an M1, so I think I should help with `Aarch64 neon supports float16x{4,8}`, but not so sure about this. However, I'm not familiar with `stdarch`, so some guidance is appreciated!

**Comment by tgross35** [comment]
Aarch64 actually got added at some point, we have the types as e.g. https://doc.rust-lang.org/nightly/core/arch/aarch64/struct.float16x4_t.html. I updated the tracking issue just now. But you don't necessarily need the hardware to develop for it; help is still welcome!

That being said, I don't actually know how much more can be done here. Arm32 SVE and RISC-V "V" both need runtime-sized types which has been a blocker for a while (there is some very unstable ongoing work to make this possible).

Probably the most helpful thing at this point would be getting [`std::simd`](https://doc.rust-lang.org/std/simd/index.html) wired up to work with these types. This should be pretty straightforward afaik, adding aliases like `f16x1` and implementing `SimdFloat`.

**Comment by LorrensP-2158466** [comment]
Alright, I'd like to help wherever I can!

Quickly looking at the impls of `f32` and `f64`, I'm guessing I need to
- add `f16` to the `alias!` macro
- implement `SimdElement` for f16
- add it to the `for_base_ops!` macro
- probably some other things where `f32` and `f64` are implemented.
- add tests

Anything I missed?

Also, do I need to make the changes in the `rust` repo or in the `portable-simd` repo?

**Comment by tgross35** [comment]
I don't really know enough about it to answer those questions. @calebzulawski mind taking a look at the above comment? And also just acking that you're okay getting portable simd interfaces started for `f16` since it's still unstable.

**Comment by calebzulawski** [comment]
The biggest lift will be updating all of the relevant intrinsics in [std::intrinsics::simd](https://doc.rust-lang.org/std/intrinsics/simd/index.html) to support f16. Once that's done, then std::simd can be updated to support f16.

**Comment by LorrensP-2158466** [comment]
Is this done in codegen?

**Comment by tgross35** [comment]
You'll probably just have to update the tests here https://github.com/rust-lang/rust/blob/36b21637e93b038453924d3c66821089e71d8baa/tests/ui/simd/intrinsic/generic-arithmetic-2.rs https://github.com/rust-lang/rust/blob/36b21637e93b038453924d3c66821089e71d8baa/tests/ui/simd/intrinsic/generic-arithmetic-pass.rs#L45 and see where things fail. From a quick search of some intrinsics, to me it looks like there's a chance that things "just work". But you might get stopped at either HIR checks or codegen lowering.

**Comment by SciMind2460** [comment]
Could we please expedite portable Simd support? IMO we can't implement proper Simd instructions for optimisation on[ libs-team#707](https://github.com/rust-lang/libs-team/issues/707) without the new instructions. Especially for `f128` as we want to implement FFI for that.
